### PR TITLE
plugins/teststeps/sshcmd: Add Timeout

### PR DESCRIPTION
**Pull request by walterchris**
Signed-off-by: Christian Walter <christian.walter@9elements.com>

### CODE REVIEW

File: plugins/teststeps/sshcmd/sshcmd.go [hunk 1]
```diff
@@ -253,6 +276,9 @@ func (ts *SSHCmd) validateAndPopulate(params test.TestStepParameters) error {
 	}
 	ts.Args = params.Get("args")
 	ts.Expect = params.GetOne("expect")
+	if params.GetOne("timeout").IsEmpty() {
+		ts.Port = test.NewParam(defaultTimeoutParameter)
```
Comment (tfg13) [inline plugins/teststeps/sshcmd/sshcmd.go:280]:
```suggestion
		ts.Timeout = test.NewParam(defaultTimeoutParameter)
```

File: plugins/teststeps/sshcmd/sshcmd.go [hunk 2]
```diff
@@ -101,6 +104,18 @@ func (ts *SSHCmd) Run(cancel, pause <-chan struct{}, ch test.TestStepChannels, p
...
+		if err != nil {
+			return fmt.Errorf("cannot expand timeout parameter: %v", err)
+		}
+
+		timeout, err := time.ParseDuration(timeoutStr)
+		if err != nil {
+			return fmt.Errorf("cannot parse timeout paramter: %v", err)
+		}
+
+		timeTimeout := time.Now().Add(timeout)
```
Comment (tfg13) [inline plugins/teststeps/sshcmd/sshcmd.go:117]: Do we really need param expansion for the timeout? If not we could do the parsing in the validate method, so we catch typos at job ingestion time.

File: plugins/teststeps/sshcmd/sshcmd.go [hunk 3]
```diff
@@ -186,31 +201,39 @@ func (ts *SSHCmd) Run(cancel, pause <-chan struct{}, ch test.TestStepChannels, p
...
+			case <-cancel:
+				return session.Signal(ssh.SIGKILL)
+			case <-pause:
+				return session.Signal(ssh.SIGKILL)
+			default:
+				if time.Now().After(timeTimeout) {
+					return fmt.Errorf("timed out after %s", timeout)
+				}
+				session.Signal(ssh.Signal("CONT"))
+				time.Sleep(250 * time.Millisecond)
```
Comment (tfg13) [inline plugins/teststeps/sshcmd/sshcmd.go:235]: You don't need to busy wait for this, I think the usual way to do this in go is:
```suggestion
		case <-time.After(timeTimeout):
				return fmt.Errorf("timed out after %s", timeout)
				session.Signal(ssh.Signal("SIGKILL"))
```
This should only ever be reached if the first case was not reached before.
(I think you also want to abort execution then with SIGKILL)

File: plugins/teststeps/sshcmd/sshcmd.go [hunk 4]
```diff
@@ -186,31 +201,39 @@ func (ts *SSHCmd) Run(cancel, pause <-chan struct{}, ch test.TestStepChannels, p
...
+				if err == nil {
+					// Execute expectations
+					expect := ts.Expect.String()
+					if expect == "" {
+						log.Warningf("no expectations specified")
 					} else {
-						return fmt.Errorf("match for %s not found for target %v", expect, target)
+						re := regexp.MustCompile(expect)
+						matches := re.FindAll(stdoutBuf.Bytes(), -1)
+						if len(matches) > 0 {
```
Comment (tfg13) [inline plugins/teststeps/sshcmd/sshcmd.go:216]: what is this `stdoutBuf` stuff doing here, I assume it leaked from one of the other PRs?

**Review by tfg13 — CHANGES_REQUESTED** [review]

File: plugins/teststeps/sshcmd/sshcmd.go (hunk 3 above)
Comment (walterchris) [reply plugins/teststeps/sshcmd/sshcmd.go:235]: I know that we do not need busy waiting for this. This actually comes from https://github.com/facebookincubator/contest/issues/152 - I had the problem that the other connection closed the connection. Turned out that I constantly need to fire Signals in order to keep the connection alive. I tried implementing keep alive which did not help - so I just this.

File: plugins/teststeps/sshcmd/sshcmd.go (hunk 4 above)
Comment (walterchris) [reply plugins/teststeps/sshcmd/sshcmd.go:216]: Yes that should not be here. Let me fix this.

File: plugins/teststeps/sshcmd/sshcmd.go (hunk 3 above)
Comment (tfg13) [reply plugins/teststeps/sshcmd/sshcmd.go:235]: Aaaah, that is what the CONT is for. Please drop a comment in then, so people know :)

Comment (tfg13) [reply plugins/teststeps/sshcmd/sshcmd.go:235]: Wait, but why does this help against the connection being closed explicitly? Sry, not sure I follow :/
And 4 signals per second seems like a lot :D

**Pull request closed without merging**